controlplane/kubeadm/controllers/scale.go

```diff
@@ -130,6 +143,93 @@ func (r *KubeadmControlPlaneReconciler) scaleDownControlPlane(
 	return ctrl.Result{Requeue: true}, nil
 }
 
+// preflightChecks checks if the control plane is stable before proceeding with a scale up/scale down operation.
+// If the controlol plane is not stable, it returns a suggested requeue.
+//
+// NOTE: this func uses KCP conditions, it is required to call reconcileControlPlaneConditions before this.
+func (r *KubeadmControlPlaneReconciler) preflightChecks(_ context.Context, controlPlane *internal.ControlPlane) (ctrl.Result, error) { //nolint:unparam
+	logger := r.Log.WithValues("namespace", controlPlane.KCP.Namespace, "kubeadmControlPlane", controlPlane.KCP.Name, "cluster", controlPlane.Cluster.Name)
+
+	// If there is no KCP-owned control-plane machines, then control-plane has not been initialized yet,
+	// so it is considered ok to proceed.
+	if controlPlane.Machines.Len() == 0 {
+		return ctrl.Result{}, nil
+	}
+
+	// If there are deleting machines, wait for the operation to complete.
+	if controlPlane.HasDeletingMachine() {
+		logger.Info("Waiting for machines to be deleted", "Machines", strings.Join(controlPlane.Machines.Filter(machinefilters.HasDeletionTimestamp).Names(), ", "))
+		return ctrl.Result{RequeueAfter: deleteRequeueAfter}, nil
+	}
+
+	// Check machine health conditions; if there are conditions with False or Unknown, then wait.
+	allMachineHealthConditions := []clusterv1.ConditionType{
+		controlplanev1.MachineAPIServerPodHealthyCondition,
+		controlplanev1.MachineControllerManagerPodHealthyCondition,
+		controlplanev1.MachineSchedulerPodHealthyCondition,
+	}
+	if controlPlane.IsEtcdManaged() {
+		allMachineHealthConditions = append(allMachineHealthConditions,
+			controlplanev1.MachineEtcdPodHealthyCondition,
+			controlplanev1.MachineEtcdMemberHealthyCondition,
+		)
+	}
+	machineErrors := []error{}
+	for _, machine := range controlPlane.Machines {
+		for _, condition := range allMachineHealthConditions {
+			if err := preflightCheckCondition(machine, condition); err != nil {
+				machineErrors = append(machineErrors, err)
+			}
+		}
+	}
+	if len(machineErrors) > 0 {
+		aggregatedError := kerrors.NewAggregate(machineErrors)
+		r.recorder.Eventf(controlPlane.KCP, corev1.EventTypeWarning, "ControlPlaneUnhealthy",
+			"Waiting for control plane to pass preflight checks to continue reconciliation: %v", aggregatedError)
+		logger.Info("Waiting for control plane to pass preflight checks", "failures", aggregatedError.Error())
+
+		return ctrl.Result{RequeueAfter: preflightFailedRequeueAfter}, nil
+	}
```

Comment:
This will result in some duplicated messages because machine problems are aggregated at KCP level as well.
So we are reporting machine error alone first without duplications and then, if there are no machine errors, we report KCP only errors (errors that can be linked to any machine)